client/js/jsonedit.js

```diff
@@ -307,24 +336,19 @@ const jeCommands = [
       if(jeMode != 'macro') {
         jeWidget = null;
         jeMode = 'macro';
-        jeSetEditorContent('// this code will be called for\n// every widget as variable w\n\n// variable v is a persistent object you\n// can use to store other information\n\nif(w.deck)\n  w.gotit = true;');
+        jeSetEditorContent(jeMacroPreset);
         jeColorize();
       } else {
         jeJSONerror = null;
         try {
           const macro = new Function(`"use strict";return (function(w, v) {${jeGetEditorContent()}})`)();
           const variableState = {};
-          for(const [ id, w ] of widgets) {
+          for(const w of [...widgets.values()]) { // shallow copy because we might create new widgets by changing the id
             const s = JSON.stringify(w.state);
             const newState = JSON.parse(s);
             macro(newState, variableState);
             batchStart();
-            if(s != JSON.stringify(newState)) {
-              for(const key in w.state)
-                if(newState[key] === undefined)
-                  newState[key] = null;
-              sendPropertyUpdate(id, newState);
-            }
+            await updateWidget(JSON.stringify(newState), s);
```

Comment:
Changing the `id` using macros led to a broken room state. `updateWidget` already takes care of that stuff and this change makes sense anyway.

Reply:
Is this still a draft?

Reply:
Oh. No. The comment doesn't mean it's broken now. It was just meant to explain why I changed it.